Jenkinsfile

```diff
@@ -92,56 +99,135 @@ pipeline {
-
-                stash name: 'authnGcpTokens', includes: 'gcp_token_valid,gcp_token_invalid_audience,gcp_token_standard_format,gcp_token_user,gcp_token_non_existing_host,gcp_token_non_existing_account,gcp_token_non_rooted_host', allowEmpty:false
-                env.GCP_TOKENS_FETCHED = "true"
+                echo '-- Google Compute Engine allocated'
+                echo '-- Get compute engine instance project name from Google metadata server.'
+                env.GCP_PROJECT = sh (
+                    script: 'curl -s -H "Metadata-Flavor: Google" \
+                    "http://metadata.google.internal/computeMetadata/v1/project/project-id"',
+                    returnStdout: true
+                ).trim()
+                unstash 'get_gce_tokens_script'
+                sh './get_gce_tokens_to_files.sh'
+                stash name: 'authnGceTokens', includes: 'gce_token_*', allowEmpty:false
               }
             }
           }
+          post {
+           failure {
+            script {
+              env.GCP_ENV_ERROR = "true"
+            }
+           }
+           success {
+            script {
+              env.GCE_TOKENS_FETCHED = "true"
+            }
+            echo '-- Finished fetching GCE tokens.'
+           }
+          }
         }
-        stage('GCP Authenticator') {
+        /**
+        * We have 3 stages for GCP Authenticator tests.
+        * In this stage, Google SDK container executes a script to deploy a function,
+        * the function accepts audience in query string and returns a token with that audience.
+        * All the tokens required for testings are obtained and written to function directory, the post stage branch
+        * deletes the function.
+        * This stage depends on stage: 'GCP Authenticator preparation - Allocate GCE Instance' to set
+        * the GCP project env var.
+        */
+        stage('GCP Authenticator preparation - Allocate Google Function') {
+          environment {
+            GCP_FETCH_TOKEN_FUNCTION = "fetch_token_${BUILD_NUMBER}"
+            IDENTITY_TOKEN_FILE = 'identity-token'
+            GCP_OWNER_SERVICE_KEY_FILE = "sa-key-file.json"
+            GCP_ZONE="us-central1"
+          }
           steps {
+            echo "Waiting for GCP project name (Set by stage: 'GCP Authenticator preparation - Allocate GCE Instance')"
             timeout(time: 10, unit: 'MINUTES') {
               waitUntil {
                 script {
-                  return (env.GCP_TOKENS_FETCHED == "true")
+                  return (env.GCP_PROJECT != null  || env.GCP_ENV_ERROR == "true")
                 }
               }
             }
             script {
-              dir('ci/authn-gcp/tokens') {
-                unstash 'authnGcpTokens'
+              if (env.GCP_ENV_ERROR == "true") {
+                error('GCP_ENV_ERROR cannot deploy function')
               }
 
+              dir('ci/authn-gcp') {
+                sh './deploy_function_and_get_tokens.sh'
```

Comment:
Do you miss here a step to stash the gcf tokens? 
Like the previous step: `stash name: 'authnGceTokens', includes: 'gce_token_*', allowEmpty:false`

Reply:
I stash because the script execute within a gce node. This is not the case here.